test/core/event_engine/test_suite/client_test.cc

```diff
@@ -12,9 +12,188 @@
+                   .ok());
+
+    // Send from server to client and verify data read at the client.
+    GPR_ASSERT(mgr.TransferFromServer(/*connection_id=*/connection_id,
+                                      /*write_data=*/GetNextSendMessage())
+                   .ok());
+  }
+  mgr.CloseConnection(connection_id);
+}
+
+// Create a N listeners and M connections where M > N and exchange and verify
+// data over each connection.
+TEST_F(EventEngineClientTest, MatrixOfConnectionsToOracleListenersTest) {
+  ConnectionManager mgr(this->NewEventEngine(), this->NewOracleEventEngine());
+  static constexpr int kStartPortNumber = 7000;
+  static constexpr int kNumListeners = 10;
+  static constexpr int kNumConnections = 100;
+  static constexpr int kMinMessagesInConnection = 10;
+  static constexpr int kMaxMessagesInConnection = 100;
+  std::vector<std::string> target_addrs;
+  std::vector<int> connections;
+  for (int i = 0; i < kNumListeners; i++) {
+    std::string target_addr =
+        absl::StrCat("ipv6:[::1]:", std::to_string(kStartPortNumber + i));
+    // Start an oracle localhost ipv6 listener
+    GPR_ASSERT(mgr.StartListener(target_addr, true).ok());
+    target_addrs.push_back(target_addr);
+  }
+  absl::SleepFor(absl::Milliseconds(500));
+  for (int i = 0; i < kNumConnections; i++) {
+    // Create a test event engine client endpoint and connect to a random oracle
+    // listener. Verify that the connection succeeds.
+    auto status = mgr.CreateConnection(
+        target_addrs[rand() % kNumListeners], absl::InfiniteFuture(),
+        [](absl::Status status) { GPR_ASSERT(status.ok()); }, false);
+    GPR_ASSERT(status.ok());
+    connections.push_back(*status);
+  }
+  std::vector<std::thread> threads;
+  threads.reserve(kNumConnections);
+  for (int i = 0; i < kNumConnections; i++) {
+    // For each connection, simulate a parallel bi-directional data transfer.
+    // All bi-directional transfers are run in parallel across all connections.
+    // Each bi-directional data transfer uses a random number of messages.
+    threads.emplace_back([&, connection_id = connections[i]]() {
+      std::random_device rd;
+      std::mt19937 gen(rd());
+      // Randomize the number of messages per connection to randomize
+      // its lifetime.
+      std::uniform_real_distribution<> dis(kMinMessagesInConnection,
+                                           kMaxMessagesInConnection);
+      int kNumExchangedMessages = dis(gen);
+      std::vector<std::thread> workers;
+      workers.reserve(2);
+      auto worker = [&mgr, connection_id,
+                     kNumExchangedMessages](bool client_to_server) {
+        for (int i = 0; i < kNumExchangedMessages; i++) {
+          // If client_to_server is true, send from client to server and verify
+          // data read at the server. Otherwise send data from server to client
+          // and verify data read at client.
+          if (client_to_server) {
+            GPR_ASSERT(
```

Comment:
You'll want to use the GoogleTest ASSERT_* and EXPECT_* macros in tests, not GPR_ASSERT.

Reply:
Done.